### PR TITLE
fix proxing live collections (close #1096)

### LINKED ISSUE #1096: Wrong proxing methods that returns live node collection
Reproduced in playrground. Relates to https://github.com/DevExpress/testcafe/issues/1442

### PR DESCRIPTION
**Pull request by miherlosev**

### CODE REVIEW

**Comment by miherlosev** [comment]
HealthMonitor run - https://testcafe-hhhm.devexpress.com/history?timestamp=1499076702502
@LavrovArtem @churkin @AlexanderMoskovkin

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 1]
```diff
@@ -327,6 +328,9 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
                     else if (domUtils.isShadowUIElement(el))
                         ShadowUI.markElementAndChildrenAsShadow(el);
 
+                    else if (!domUtils.isShadowUIElement(el))
```
Comment (churkin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Very strange condition

File: src/client/sandbox/node/element.js [hunk 2]
```diff
@@ -577,14 +578,20 @@ export default class ElementSandbox extends SandboxBase {
...
-    onElementRemoved (el) {
+    _onElementRemoved (el) {
         if (domUtils.isBodyElement(el))
             this.shadowUI.onBodyElementMutation();
 
         else if (domUtils.isBaseElement(el))
             urlResolver.updateBase(getDestLocation(), this.document);
+
+        if (el.tagName && !domUtils.isShadowUIElement(el))
+            this.liveNodeListFactory.onElementMutation(el);
```
Comment (churkin) [inline src/client/sandbox/node/element.js]: I think `el.parentNode` has been mutated, not el

File: src/client/sandbox/node/live-node-list/factory.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import {
+    NodeListPropertiesDecorator,
+    HTML_COLLECTION_PROPERTIES_DECORATOR,
+    NODE_LIST_PROPERTIES_DECORATOR
+} from './property-decorator';
+import TagCache from './tag-cache';
+import { getTagName } from '../../../utils/dom';
+import nativeMethods from '../../native-methods';
+
+var arrayForEach = Array.prototype.forEach;
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/factory.js]: consts

File: src/client/sandbox/node/live-node-list/property-decorator.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+const DEFINED_PROPERTIES_COUNT = 10000;
+
+export class NodeListPropertiesDecorator {
+    constructor () {
+        var defineProperty = function (index, isEnumerable) {
+            Object.defineProperty(this, index, {
+                enumerable:   isEnumerable,
+                configurable: true,
+                get:          function () {
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: Use new line here

File: src/client/sandbox/node/live-node-list/property-decorator.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                }
+            });
+        };
+
+        for (var i = 0; i < DEFINED_PROPERTIES_COUNT; i++)
+            defineProperty.call(this, i, false);
+
+        Object.defineProperty(this, '_defineProperty', { value: defineProperty });
+        Object.defineProperty(this, '_refreshNodeList', {
+            value: () => {
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: Extra `{,}`?

File: src/client/sandbox/node/live-node-list/tag-cache.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+export default class TagCache {
+    constructor () {
+        this._cache = [];
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/tag-cache.js]: Use hash table instead of array

File: src/client/sandbox/node/element.js (hunk 2 above)
Comment (miherlosev) [inline src/client/sandbox/node/element.js]: `onElementMutation` looks like bad name.
`onElementAddedOrRemoved` is more cleaner.

File: src/client/sandbox/node/live-node-list/property-decorator.js (hunk 5 above)
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: No. You cannot define arrow function with `throw` statement without brackets.

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

File: src/client/sandbox/node/live-node-list/factory.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        nativeMethods.addEventListener.call(document, 'DOMContentLoaded', () => {
+            this.domContentLoadedEventRaised = true;
+
+            LiveNodeListFactory._notifyWrappersAboutDOMContentLoadedEvent(this.wrappersWithAsteriskTag);
+            LiveNodeListFactory._notifyWrappersAboutDOMContentLoadedEvent(this.wrappersWithReqularTag);
+        });
+    }
+
+    static _markWrappersAsDirty (wrappers) {
+        arrayForEach.call(wrappers, wrapper => {
```
Comment (LavrovArtem) [inline src/client/sandbox/node/live-node-list/factory.js]: May be `for (var wrapper of wrappers)`

**Comment by LavrovArtem** [comment]
Also pay attention to the `replaceChild` function.

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR
@churkin @ale

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR

**Comment by miherlosev** [comment]
ping @churkin @AlexanderMoskovkin

File: src/client/sandbox/node/live-node-list/factory.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        else {
+            // NOTE: For iframe's collections
+            NodeListPropertiesDecorator.prototype = nodeList;
+            LiveNodeListWrapper.prototype         = new NodeListPropertiesDecorator();
+        }
+
+        return new LiveNodeListWrapper(nodeList, domContentLoadedEventRaised, tagName);
+    }
+
+    createNodeListForGetElementsByTagNameFn ({ nodeList, tagName }) {
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/factory.js]: Why we need so difficult name if this class doesn't have any other creators?

File: src/client/sandbox/node/live-node-list/factory.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    HTML_COLLECTION_PROPERTIES_DECORATOR,
+    NODE_LIST_PROPERTIES_DECORATOR
+} from './property-decorator';
+import TagCache from './tag-cache';
+import { getTagName, isShadowUIElement } from '../../../utils/dom';
+import nativeMethods from '../../native-methods';
+
+const arrayFilter = Array.prototype.filter;
+
+export default class LiveNodeListFactory {
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/factory.js]: I think this is not a factory

File: src/client/sandbox/node/live-node-list/factory.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        this.wrappersWithReqularTag      = [];
+        this.wrappersWithAsteriskTag     = [];
+        this.domContentLoadedEventRaised = false;
+        this.tagCache                    = new TagCache();
+
+        nativeMethods.addEventListener.call(document, 'DOMContentLoaded', () => {
+            this.domContentLoadedEventRaised = true;
+
+            LiveNodeListFactory._notifyWrappersAboutDOMContentLoadedEvent(this.wrappersWithAsteriskTag);
+            LiveNodeListFactory._notifyWrappersAboutDOMContentLoadedEvent(this.wrappersWithReqularTag);
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/factory.js]: We can combine these calls

File: src/client/sandbox/node/live-node-list/property-decorator.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+const DEFINED_PROPERTIES_COUNT = 10000;
+
+export class NodeListPropertiesDecorator {
```
Comment (churkin) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: I don't like `Decorator` term in this case

File: src/client/sandbox/node/live-node-list/factory.js (hunk 9 above)
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/factory.js]: It's a factory class that creates live node list wrappers

File: src/client/sandbox/node/live-node-list/factory.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+import { getTagName, isShadowUIElement } from '../../../utils/dom';
+import nativeMethods from '../../native-methods';
+
+const arrayFilter = Array.prototype.filter;
+
+export default class LiveNodeListFactory {
+    constructor () {
+        // NOTE: Now there is an implementation only for the 'getElementsByTagName' function
+        // Later we will have separate parts for all appropriate functions (getElementsByClassName, getElementsByName)
+        this.wrappersWithReqularTag      = [];
```
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/factory.js]: Yes

File: src/client/sandbox/node/live-node-list/factory.js (hunk 10 above)
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/factory.js]: Ok

File: src/client/sandbox/node/live-node-list/factory.js (hunk 8 above)
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/factory.js]: https://github.com/DevExpress/testcafe-hammerhead/pull/1167/files#diff-c1e7400e5cb4d585b63684945efc511cR15

File: src/client/sandbox/node/live-node-list/property-decorator.js (hunk 11 above)
Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: I propose `LiveNodeListWrapperBase`, `HTML_COLLECTION_WRAPPER`, `NODE_LIST_WRAPPER`.

Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/property-decorator.js]: What do think about it?

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR

File: src/client/sandbox/node/live-node-list/wrapper-base.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            value: () => {
+                throw new Error('Not implemented');
+            }
+        });
+    }
+}
+
+LiveNodeListWrapperBase.prototype = NodeList.prototype;
+
+export const NODE_LIST_WRAPPER = new LiveNodeListWrapperBase();
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/node/live-node-list/wrapper-base.js]: usually we use UPPER_CASE for string/numeric consts or objects with srting/numeric consts. But here is just an instance of the class. I expect to see here camelcase as usually

Comment (miherlosev) [inline src/client/sandbox/node/live-node-list/wrapper-base.js]: > usually we use UPPER_CASE for string/numeric consts or objects with string/numeric consts.

Not only. We use upper case notation for object of any type that uses as constant or as etalon object.
For example, see https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/utils/dom.js#L32

`NODE_LIST_WRAPPER`, `HTML_COLLECTION_WRAPPER` are etalons for construct live node list collection of an appropriate type.

Comment (AlexanderMoskovkin) [inline src/client/sandbox/node/live-node-list/wrapper-base.js]: [here](https://google.github.io/styleguide/jsguide.html#naming-constant-names) is nice explanation of constants I think. 

> Before choosing constant case, consider whether the field really feels like a deeply immutable constant. For example, if any of that instance's observable state can change, it is almost certainly not a constant. Merely intending to never mutate the object is generally not enough.

I think an instance of the wrapper base class doesn't match this conditions

**Comment by miherlosev** [comment]
I will merge only after health monitor run

**Comment by miherlosev** [comment]
All works as well
https://testcafe-hhhm.devexpress.com/history?timestamp=1499947445208

**Pull request merged**